### DISCUSSION TITLE
Improve selector architecture to allow for different object types and sources

### DISCUSSION THREAD

**Discussion by wimvelzeboer** [wimvelzeboer/fflib-apex-common#10] [Ideas]
The selector should be redesigned so that it is not coupled to a database. In this new architecture the selector will retrieve data from wherever it is stored. It should not be the concern of a service class calling the selector to know where the data is coming from.
This new architecture should allow for getting the data from;

In memory from previous queries (Identity Map Design Pattern)
A Platform Cache
The Database
External WebService
The queries should also be able to roll-over. Like when the requested record is not in memory, then the next selector in line should be called as the record might be in the Platform cache. If it is still not found the record can be queried from a database.

**Comment by klewis-sfdc** [comment]
Would you plan to modify the existing Selector classes (and keep the "Selector" nomenclature), or implement some type of a Repository that uses Selectors when accessing the database?

**Comment by wimvelzeboer** [comment]
Hi @kenlewis-sfdo,

Yes! Now that the AEP repo has accepted the new domain structure I plan to move forward with this selector.
In my opinion a selector that fetches data from only one source (the database) is very limited. 

When I write a service method and I need a data set at some point, I am not interested in where this data is coming from, I just want the data.
So, I do not care it it comes from a database, in memory (e.g. Identity Pattern) or from an external source via a e.g. webservice call.

I will probably do something similar as I did with the Domain, and introduce a new interface named fflib_ISelector, and that the existing fflib_SObjectSelector is just one of the implementations.

If you want to contribute with this effort I am happy to discuss this further with you!

**Reply by klewis-sfdc** [reply]
I'd love to help if I can!  I was eagerly anticipating the acceptance of the new domain structure (thank you!) and am excited about all of the new architecture and capabilities you are implementing.  I have been thinking a bit about this idea and would love to see it in action.  I am curious about sharing and access checks when retrieving Salesforce data that has previously been cached.  Would the new Selector implementation re-run access checks using fflib_SecurityUtils before returning SObject data from cache perhaps?  And what about sharing rules?  Would it be possible that a User incorrectly gains access to a record that was cached by a different User?
